### ISSUE TITLE
Support Joins for Finding Values

### ISSUE THREAD

**Issue by spficklin** [tripal/tripal#1719]
### Task Description

Our current approach in fields for modeling the joins needed for inserting, updating and selecting all field values when the primary key of a base table record is known works well. However, this approach is insufficient for finding records when values in non-base tables are needed to find records in base tables. 

## Join Use Cases
Currently we support joins in fields using two different methods:
1.	CASE 1: The `store_link` action.   A join is specified using the `left_table`, `left_table_id`, `right_table` and `right_table_id` property storage settings.   The `ChadoStorage` plugin uses these values to create a “path” that is passed to the `ChadoStorage::addChadoRecordJoins()` function.
2.	CASE 2: The `read_value` action.  A join is specified using the `path` property storage settings. This path is passed directly to the `ChadoStorage::addChaodRecordJoins()` function.

## Review of how joins are currently handled.
Both CASE 1 and CASE 2 end up using the `ChadoStorage::addChadoRecordJoins()` to add joins. The code from that function that adds the join to the records object is as follows:

https://github.com/tripal/tripal/blob/6779fccbf67d58950b6186eb9c82f990d48d33bd/tripal_chado/src/Plugin/TripalStorage/ChadoStorage.php#L1714-L1727 

Note that aliases for tables in a join are given a numeric suffix based on the order in which a table is added.  If a table is only added once to a join then the suffix is always 0. But the second time, it will have a suffix of 1.  This can happen for tables such as `featureloc` because it has a `feature_id` and `srcfeature_id` column that both join to the `feature` table.  

For insert, update or selecting of a single record these aliases don’t matter because we also give an alias to the column we want via the `as` property in the `read_value` action.  So regardless of the table aliasing we can always pull out a unique column. 

Also, note that the “base” table in the joins is referred to as “ct” (an abbreviation for “chado table”, or “base table”).

## Challenges for finding values.
For finding values, we intend to find all base table records that match a set of criteria.  Theoretically, we should be able to find records using values set for any field property on the content type.  

### Publishing Use Case:
One use case for finding values, is in publishing. For example, the “Genome Assembly” content type requires that the `type_id` for an analysis base record be set in the `analysisprop` table.  So, to publish, we must first find all the analyses with an associated record in the analysisprop table.   If we look at the `ChadoAdditionalTypeTypeDefault` plugin class. We see that the join is specified via a property with a `store_link` action:
https://github.com/tripal/tripal/blob/6779fccbf67d58950b6186eb9c82f990d48d33bd/tripal_chado/src/Plugin/Field/FieldType/ChadoAdditionalTypeTypeDefault.php#L132-L136
The value of the `type_id` comes from a property with a `store` action.
https://github.com/tripal/tripal/blob/6779fccbf67d58950b6186eb9c82f990d48d33bd/tripal_chado/src/Plugin/Field/FieldType/ChadoAdditionalTypeTypeDefault.php#L145-L150 

#### Problem  1
First, the `link_id` field is missing the `left_table`, `left_table_id`, `right_table` and `right_table_id` properties. That’s okay, it just hasn’t been updated yet.  But let’s assume they are there to indicate how the table should join to the base table.  I’ve fixed this in my current working branch `tv4g2-issue1515-publish_chado_content`

#### Problem 2
The record array structure that ChadoStorage used for representing how to query cannot currently handle adding conditions on joined tables.  Here’s what it current looks like (in YAML format) for finding analyses:

```
analysis:
- conditions:
    analysis_id:
      value:
      - REPLACE_BASE_RECORD_ID
      - analysis
      operation: =
  fields:
    analysis_id: ~
    name: ~
  joins:
    analysisprop:
      analysis_id:
        "on":
          left_table: analysis
          left_col: analysis_id
          right_table: analysisprop
          right_col: analysis_id
          left_alias: analysis
          right_alias: analysisprop
        columns:
        - - analysis_id
          - link_id
          - genome_assembly_type
          - link_id
```
In order to filter (or find) base records that have conditions in joined tables we need to specify the table. In my current working branch this is how I changed the array:

```
analysis:
- conditions:
    analysis:
      analysis_id:
        value:
        - REPLACE_BASE_RECORD_ID
        - analysis
        operation: =
    analysisprop:
      type_id:
        value: "31"
        operation: =
  fields:
    analysis_id: ~
    name: ~
  joins:
    analysisprop:
      analysis_id:
        "on":
          left_table: analysis
          left_col: analysis_id
          right_table: analysisprop
          right_col: analysis_id
          left_alias: analysis
          right_alias: analysisprop
        columns:
        - - analysis_id
          - link_id
          - genome_assembly_type
          - link_id
```
Notice how the table name for the condition is now above the criteria.  This fix allows us to find analyses that are only “Genome Assembly” records and publish those!  It’s currently working using this approach.

#### Problem 3
The change to the records array structure to support conditions on non-base tables works because I also had to make a change in the `ChadoStorage::addChadoRecordJoins()` function.  In this use case, the `analysisprop` table is joined to the `analysis` table because the `link_id` property provides that join. The `type_id` property takes advantage of that join.  In the current `addChadoRecordJoins()` function (see code above) the join will create the alias `janalysisprop0` for the anaysisprop table.   Unfortunately, the `type_id` property doesn’t know what the alias is.   It would be unreasonable to expect the field developer to know this alias, and even less to know the order in which tables with the same name are joined. As long is there is only one instance of a table joined all is fine if the developer knows to use the fromat similar to `janalysisporp0` in a `chado_table_alias` property setting.  Since that did not seem reasonable, I changed the `ChadoStorage::addChadoRecordJoins()` to not add the `j` prefix nor the numeric suffix if the table was the first one.   The developer need not provide any additional settings. 

### Future Issues
In the process of adding fixes for the use case above I realized that we would continue to have problems when finding values because:

1.	Field properties rely on other properties to specify a join.  This is ambiguous.   For example, If we wanted to find a feature (via the `featureloc` table) with a given name (gene) located on another feature of a given name (e.g. chromosome) then we have to join twice on the feature table.  There is no way for the field properties to know which join of the feature table they should use to set a conditional value.  The `store_link` property could set an alias, but the feature `name` property doesn’t know which alias to use.  
2.	The `addChadoRecordJoins()` sets the table aliases.  This means it’s not possible for a field to know what the table alias is.  My solution for the publishing use case works as long as the table only occurs once in the joins.

### Possible Solution
Use a specific `path` attribute for all properties.  This would follow the same convention as the `path` setting does for the properties with a `read_value` action. One change is to add the column where the value comes from to the end of the path. Consider this for the `ChadoAdditionlaTypeTypeDefault` field

```php
if ($type_table != $base_table) {
      $type_table_def = $schema->getTableDef($type_table, ['format' => 'Drupal']);
      $type_pkey_col = $type_table_def['primary key'];
      $type_fkey_col = array_keys($type_table_def['foreign keys'][$base_table]['columns'])[0];
      $link_term = $mapping->getColumnTermId($type_table, $type_fkey_col);
      $value_term = $mapping->getColumnTermId($type_table, 'value');

      $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'prop_id', $record_id_term, [
        'action' => 'store_pkey',
        'drupal_store' => TRUE,
        'path' => $type_table . "." . $type_pkey_col,
        # 'chado_table' => $type_table,
        # 'chado_column' => $type_pkey_col,
      ]);
      $properties[] =  new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'link_id', $link_term, [
        'action' => 'store_link',
        'drupal_store' => TRUE,
        'path' => $base_table . '.' . $base_pkey_col . '>' . $type_table . '.' . $type_fkey_col,
        # 'left_table' => $base_table,
        # 'left_table_id' => $base_pkey_col,
        # 'right_table' => $type_table,
        # 'right_table_id' => $type_fkey_col,
      ]);
      $properties[] =  new ChadoTextStoragePropertyType($entity_type_id, self::$id, 'value', $value_term, [
        'action' => 'store',
        'path' => $type_table . '.value',
        # 'chado_table' => $type_table,
        # 'chado_column' => 'value',
      ]);
    }

    // We need to store the numeric cvterm ID for this field.
    $properties[] = new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'type_id', $type_id_term, [
      'action' => 'store',
      'drupal_store' => TRUE,
      'path' => $base_table . '.' . $base_pkey_col . '>' . $type_table . '.' . $type_fkey_col . ';' . $type_column,,
      # 'chado_table' => $type_table,
      # 'chado_column' => $type_column,
      'empty_value' => 0
    ]);
    // This field needs the term name, idspace and accession for proper
    // display of the type.
    $properties[] = new ChadoVarCharStoragePropertyType($entity_type_id, self::$id, 'term_name', $name_term, 128, [
      'action' => 'read_value',
      'path' => $type_table . '.' . $type_column . '>cvterm.cvterm_id;name',
      # 'chado_column' => 'name',
      'as' => 'term_name'
    ]);
    $properties[] = new ChadoVarCharStoragePropertyType($entity_type_id, self::$id, 'id_space', $idspace_term, 128, [
      'action' => 'read_value',
      'path' => $type_table . '.' . $type_column . '>cvterm.cvterm_id;cvterm.dbxref_id>dbxref.dbxref_id;dbxref.db_id>db.db_id;name',
      #'chado_column' => 'name',
      'as' => 'idSpace'
    ]);
    $properties[] = new ChadoVarCharStoragePropertyType($entity_type_id, self::$id, 'accession', $accession_term, 128, [
      'action' => 'read_value',
      'path' => $type_table. '.' . $type_column . '>cvterm.cvterm_id;cvterm.dbxref_id>dbxref.dbxref_id;accession',
      # 'chado_column' => 'accession',
      'as' => 'accession'
    ]);
```
Notice that the `type_id` property fully specifies where the value comes from... as well as all of the fields There is no ambiguity as to where it's value comes from.  Since the `type_id` and `link_id` paths are the same (other than the last column name), then it's clear that the vlaues from those two properties comes from the same join.

**Comment by spficklin** [comment]
Linking to issue #1703 as that issue is derived from a similar problem resulting from our current design that I point out above

**Comment by laceysanderson** [comment]
Thanks for the write-up Stephen! It was very helpful for me to understand the issue!

I really like the consistency added by using the path in all actions now that I see your example!

I also like the change to the join array suggested for the solution to problem 2: specifically, keying each join condition by the table. 

Some things we will want to be extra careful of:
1. When the path is the same but the join cannot be shared due to different conditions (e.g. property field).
2. Ensuring that joins are added to the query in the order they are specified in the path. This should resolve issue #1703.
3. Ensuring that some joins are shared across fields for performance.
4. Ensuring we do not break the existing tests... the ONLY change needed to existing tests should be switching the property type definitions to use path.

**Note: All the above should be ok with the solution suggested by Stephen except for No. 1. See the possible extension to Stephens solution at the bottom.**

## When the path is the same but the join cannot be shared due to different conditions.

The current solution by @spficklin uses the table name to key the join conditions and the intent is that all actions with the same path will use the same set of joins.

However, there are multiple instances where the normalization of chado causes us to write multiple joins to the same table with different conditions. In some cases, the approach of applying alias' based on the order the tables are added works such as the situation with feature locations linking back to the feature table 2X (one for the feature_id and again for the srcfeature_id). However, I worry there are many examples where order alone is not sufficient.

One such example is the property field. The property types will always have a path of `base_table.base_pkey>prop_table.prop_fkey*`; however, different instances of the property field need different conditions on their join and thus cannot share a join group. 

Currently we use `chado_table_alias`/`right_table_alias`/`left_table_alias` where the alias includes the cvterm_id of the type each field is focused on. However, the solution suggested by Stephen isn't clear as to where such table alias' would be applied.

## Ensuring that joins are added to the query in the order they are specified in the path.

Use of the path should make solving this one pretty easy as we have a very clear notation for the order the joins should be added in. That said, we will want to double check that the changes to `ChadoStorage::addChadoRecordJoins()` do add the joins in the order specified in the path.

## Ensuring that some joins are shared across fields for performance.

An example of where we want joins to be shared is between the type and relationship fields. Specifically, the relationship field in Tripal 3 listed the type of the chado record on each side of the relationship. Since one side of the relationship will often be the base table, it would be ideal to share the base.type_id>cvterm.cvterm_id join between these two fields.

With the concerns raised in "the path is the same but the join cannot be shared due to different conditions", there can be a tendency to play it safe by just aliasing all joins with the field name. However, with chado being such a web we will definitely want to share joins where we can. This is just something to keep in mind so we're not too cautious.

## Ensuring we do not break the existing tests.

ChadoStorage is a beast with many interconnected parts. We want to be really careful that any of the changes suggested in this issue are in their own branch and that all the existing tests pass throughout the process! The ONLY change needed should be to update the property type definitions to include the `path`.

# Possible Solution incorporating table alias'

Essentially, we can use the table alias in the path if there is one and then define a mapping within each action using an alias in the path. The mapping would allow multiple tables to be aliased if needed for longer read_value joins. 

For example, in the property field this would look like:

```php
    return [
      new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'record_id', $record_id_term, [
        'action' => 'store_id',
        'drupal_store' => TRUE,
        'path' => $base_table . '.' . $base_pkey_col,
        # 'chado_table' => $base_table,
        # 'chado_column' => $base_pkey_col
      ]),
      new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'prop_id', $record_id_term, [
        'action' => 'store_pkey',
        'drupal_store' => TRUE,
        'path' => $base_table . '.' . $base_pkey_col . '>' . $table_alias . '.' . $prop_fk_col . ';' . $prop_pkey_col,
        'table_alias_mapping' => [
          $table_alias => $prop_table,
        ],
        # 'chado_table' => $prop_table,
        # 'chado_column' => $prop_pkey_col,
        # 'chado_table_alias' => $table_alias,
      ]),
      new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'linker_id',  $link_term, [
        'action' => 'store_link',
        'path' => $base_table . '.' . $base_pkey_col . '>' . $table_alias . '.' . $prop_fk_col,
        'table_alias_mapping' => [
          $table_alias => $prop_table,
        ],
        # 'left_table' => $base_table,
        # 'left_table_id' => $base_pkey_col,
        # 'right_table' => $prop_table,
        # 'right_table_alias' => $table_alias,
        # 'right_table_id' => $prop_fk_col,
      ]),
      new ChadoTextStoragePropertyType($entity_type_id, self::$id, 'value', $value_term, [
        'action' => 'store',
        'path' => $base_table . '.' . $base_pkey_col . '>' . $table_alias . '.' . $prop_fk_col . ';value',
        'table_alias_mapping' => [
          $table_alias => $prop_table,
        ],
        # 'chado_table' => $prop_table,
        # 'chado_table_alias' => $table_alias,
        # 'chado_column' => 'value',
        'delete_if_empty' => TRUE,
        'empty_value' => ''
      ]),
      new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'rank', $rank_term,  [
        'action' => 'store',
        'path' => $base_table . '.' . $base_pkey_col . '>' . $table_alias . '.' . $prop_fk_col . ';rank',
        'table_alias_mapping' => [
          $table_alias => $prop_table,
        ],
        # 'chado_table' => $prop_table,
        # 'chado_table_alias' => $table_alias,
        # 'chado_column' => 'rank'
      ]),
      new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'type_id', $type_id_term, [
        'action' => 'store',
        'path' => $base_table . '.' . $base_pkey_col . '>' . $table_alias . '.' . $prop_fk_col . ';type_id',
        'table_alias_mapping' => [
          $table_alias => $prop_table,
        ],
        # 'chado_table' => $prop_table,
        # 'chado_table_alias' => $table_alias,
        # 'chado_column' => 'type_id'
      ]),
    ];
```

Then the table key in the buildChadoRecords join conditions array would be the table alias. The mapping would be saved in [`$table_alias_mapping` and `$reverse_alias_mapping`](https://github.com/tripal/tripal/blob/4.x/tripal_chado/src/Plugin/TripalStorage/ChadoStorage.php#L72-L90) just as is currently done. We would also make sure there is an entry in these arrays for all tables mentioned in the `path`: either using the alias mentioned in the property type table alias mapping or adding a table_name => table_name mapping showing the alias is the same as the table name.

**Comment by laceysanderson** [comment]
I suggest we write automated tests that highlight the "things to be extra careful of" mentioned in the previous comment before doing any development. I am happy to start this in the branch attached to this issue: `tv4g1-issue1719-usePathForAllActions`.

@spficklin if you don't see any issues with my extension of your solution to incorporate alias' then feel free to start these fixes in `tv4g1-issue1719-usePathForAllActions` branch. 

In order to avoid merge conflicts and focus us each where we excel, I propose that 
- I only make changes in the tests: add the ones I suggested and update the existing ones to include a path and mapping. I won't remove the table/column notation in the tests that we currently use but instead just add the path/alias to it. That way the tests should not break at any point during your development.
- You focus on the changes needed in chado storage and updating the existing fields.

**Comment by spficklin** [comment]
Yes, I like all of your comments suggested changes.   I'll work on implementing my part.  After some thought, I still see issues:

1.  Minor: The path strings are kind of long and ugly. I find myself having to remind myself where the different symbols get used.  
2.  Minor: It seems awkward to have two variables to specify how joins work (e.g. `path` and `table_alias_mapping`.
3.  Major:  If we use the `cvtermprop` in any field then the `table_alias_mapping` won't be specific enough.
4.  Potential problem:  Would there ever be a case where we need to specify the type of join?  Maybe a field wants to be flexible....?
5. Potential problem:  Would a field ever want to join deeply to ensure a criteria is met but use a value from a different table in the join?

I think we could solve all these issues if we did the following:

1. Convert the `path` string to a nested array where joins are nested.  Each join array can include the type of join (e.g. 'inner', 'outer', etc.), the table to join on, the left and right columns and the alias.
2. To distinguish which column in all of the tables has the value, we use a `value_column` setting.  That way someone could still do a join past the table with the value (to ensure a fixed criteria) but use the value from a table in an earlier join.
3. Replace `as` with `value_alias` to be consistent, and this can get used anywhere the `value_column` is used.

Here are some examples:

Example: Property field

```php
      // Field that reads a value directly from the base table: 
      new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'record_id', $record_id_term, [
        'action' => 'store_id',
        'drupal_store' => TRUE,
        'path' => [
          'chado_table' => $base_table,
          'table_alias' => $base_table,  // optional if not needed
          'value_column' => $base_pkey_col,
          'value_alias' => $base_pkey_col  // optional if not needed
        ],
        # 'chado_table' => $base_table,
        # 'chado_column' => $base_pkey_col
      ]),

      // Field that uses a join to get a value:
      new ChadoIntStoragePropertyType($entity_type_id, self::$id, 'prop_id', $record_id_term, [
        'action' => 'store_pkey',
        'drupal_store' => TRUE,
        'path' => [
          'chado_table' => $base_table,
          'table_alias' => $base_table,  // optional if not needed
          'join' => [
            'type' => 'inner',  // defaults to 'inner' if not specified.
            'chado_table' => $prop_table,
            'table_alias' => $table_alias, // optional if not needed
            'left_column' =>  $base_pkey_col, 
            'right_column' => $prop_fk_col,
            'value_column' => $prop_pkey_col,
            'value_alias' => $prop_pkey_col // optional if not needed
          ],
        ],
        # 'path' => $base_table . '.' . $base_pkey_col . '>' . $table_alias . '.' . $prop_fk_col . ';' . $prop_pkey_col,
        # 'chado_table' => $prop_table,
        # 'chado_column' => $prop_pkey_col,
        # 'chado_table_alias' => $table_alias,
      ]),

```

Another example:  additional type field:
```php

    // Field that has multiple joins to get a value.
    $properties[] = new ChadoVarCharStoragePropertyType($entity_type_id, self::$id, 'id_space', $idspace_term, 128, [
      'action' => 'read_value',
      // 'path' => $type_table . '.' . $type_column . '>cvterm.cvterm_id;cvterm.dbxref_id>dbxref.dbxref_id;dbxref.db_id>db.db_id;name',
      'path' => [
          'chado_table' => 'cvterm',
          'table_alias' => 'cvterm',  // optional if not needed
          'join' => [
            'type' => 'inner',  // defaults to 'inner' if not specified.
            'chado_table' => 'cvterm',
            'table_alias' => 'cvterm',  // optional if not needed
            'left_column' => $type_column,
            'right_column' => 'cvterm_id,
            'join' => [
              'type' => 'inner',  // defaults to 'inner' if not specified.
              'chado_table => 'dbxref',
              'table_alias' => 'dbxref', // optional if not needed
              'left_columns' => 'dbxref_id',
              'right_columns' => 'dbxref_id',
              'join' => [
                'type' => 'inner',  // defaults to 'inner' if not specified.
                'chado_table' => 'db',
                'table_alias' => 'db', // optional if not needed
                'left_column' => 'db_id',
                'right_column' => 'db_id'
                'value_column'  => 'name'
                'value_alias' => 'idSpace' // optional if not needed
              ],
            ],
          ],
        ], 
      #'chado_column' => 'name',
      'as' => 'idSpace'
    ]);

```

**Comment by laceysanderson** [comment]
I have a couple of questions but I'll ask one at a time.

> Major: If we use the cvtermprop in any field then the table_alias_mapping won't be specific enough.

I don't see why table_alias_mapping is not specific enough here. The mapping is `alias => table` and thus can support linking to the same table or even in cycles. 

When I think through the example of having annotations for features stored as cvterms and the evidence codes for these associations in the cvtermprop, I end up with the following structure to the field where each bullet point is a property type.
- Display the association term name.
  - path: feature.feature_id>feature_cvterm.feature_id;feature_cvterm.cvterm_id>cvterm.cvterm_id;name
  - no table alias needed.
 
**Display the value of evidence code with a specific type (e.g. cvtermprop.type_id=1234)**
- Restrict the type
  - path: feature.feature_id>feature_cvterm.feature_id;feature_cvterm.cvterm_id>cvterm.cvterm_id;cvterm.cvterm_id>cvtermprop_1234.cvterm_id;type_id
  - fixed_value: 1234
  - table_alias_mapping: cvtermprop_1234 => cvtermprop
- Grab the value
  - path: feature.feature_id>feature_cvterm.feature_id;feature_cvterm.cvterm_id>cvterm.cvterm_id;cvterm.cvterm_id>cvtermprop_1234.cvterm_id;value
  - table_alias_mapping: cvtermprop_1234 => cvtermprop
  
Can you explain the problem here with an example maybe? Also, what would this look like if we went with your suggested array structure?

**Comment by laceysanderson** [comment]
> Potential problem: Would there ever be a case where we need to specify the type of join? Maybe a field wants to be flexible....?

Perhaps but I wonder if it might be better to keep the field property types simpler and if a field does need this they can handle it themselves either in the widget or using the function action type.

> Potential problem: Would a field ever want to join deeply to ensure a criteria is met but use a value from a different table in the join?

Perhaps but in this case they would have one property type joining deeply and a separate property type only joining to the table where they want to grab the value. Our current setup should handle this just fine.

**Comment by spficklin** [comment]
> I don't see why table_alias_mapping is not specific enough here. The mapping is alias => table and thus can support linking to the same table or even in cycles.

I didn't realize the alias was being used in the path string.  I see now that it is.   So you're right it would be specific and could work..  

> When I think through the example of having annotations for features  ... 

I think the example you gave isn't quite right.  The evidence codes would be in the `feature_cvtermprop` table... not the `cvtermprop` table... right? The `cvtermprop` table is for properties of a cvterm.... which it could be possible that someone would want to filter on (e.g. all features assigned a cvterm that has a given property type).  But, if aliases are are in the path string then I think it would be fine. 

Some thoughts on comments related to the possible issues: 

>Potential problem: Would there ever be a case where we need to specify the type of join? Maybe a field wants to be flexible....?
>> Perhaps but I wonder if it might be better to keep the field property types simpler and if a field does need this they can handle it themselves either in the widget or using the function action type.

For finding values in a search you don't have the benefit of the widget.  The function action may be helpful but if you could easily set it up in the field why make someone write a function?  

> Potential problem: Would a field ever want to join deeply to ensure a criteria is met but use a value from a different table in the join?
>> Perhaps but in this case they would have one property type joining deeply and a separate property type only joining to the table where they want to grab the value. Our current setup should handle this just fine.

The properties are meant to provide values for the field... either values that get displayed to the user or values that need to be stored for finding the records in Chado.  Using a property to force a join so that other properties can find what they need doesn't seem right to me.

**Comment by spficklin** [comment]
After a brief discussion with @laceysanderson on Slack, we're going to go with her first suggestion for the sake of time constraints and revisit the design suggestion I made later.

The branch tv4g1-issue1719-usePathForAllActions is now created.  I'm working on those fixes now.

**Comment by laceysanderson** [comment]
I have updated all the tests in the tripal_chado/tests/src/Kernel/Plugin/ChadoStorage directory to use yaml files to define the field property type definitions. 

I've also updated all the property type definitions in these tests to include a path and table_alias_mapping as designed above. 

The only ones I was not able to update are those with the `read_value` action which already have a path. In these cases I added `#READY ;column_name` to the end of each path in preparation so once that is implemented you simply need to do a find and replace to remove the `#READY` part.

All these changes are in the tv4g1-issue1719-usePathForAllActions branch.
